src/L1/rollup/ScrollChain.sol

```diff
@@ -436,76 +506,94 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
+                    _dataHash,
+                    _blobDataProof[0:64],
+                    _blobVersionedHash
+                )
             );
-            // We verify that the point evaluation precompile call was successful by testing the latter 32 bytes of the
-            // response is equal to BLS_MODULUS as defined in https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile
-            if (!success) revert ErrorCallPointEvaluationPrecompileFailed();
-            (, uint256 result) = abi.decode(data, (uint256, uint256));
-            if (result != BLS_MODULUS) revert ErrorUnexpectedPointEvaluationPrecompileOutput();
         }
 
-        // verify previous state root.
-        if (finalizedStateRoots[_batchIndex - 1] != _prevStateRoot) revert ErrorIncorrectPreviousStateRoot();
+        // load version from batch header, it is always the first byte.
+        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
+        // verify batch
+        IRollupVerifier(verifier).verifyAggregateProof(batchVersion, _batchIndex, _aggrProof, _publicInputHash);
+
+        // Pop finalized and non-skipped message from L1MessageQueue.
+        uint256 _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
+        _popL1MessagesMemory(
+            BatchHeaderV1Codec.getSkippedBitmapPtr(batchPtr),
+            _totalL1MessagesPoppedOverall,
+            BatchHeaderV0Codec.getL1MessagePopped(batchPtr)
+        );
 
-        // avoid duplicated verification
-        if (finalizedStateRoots[_batchIndex] != bytes32(0)) revert ErrorBatchIsAlreadyVerified();
+        _afterFinalizeBatch(_totalL1MessagesPoppedOverall, _batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
+    }
 
-        // compute public input hash
-        bytes32 _publicInputHash = keccak256(
-            abi.encodePacked(
-                layer2ChainId,
-                _prevStateRoot,
-                _postStateRoot,
-                _withdrawRoot,
-                _dataHash,
-                _blobDataProof[0:64],
-                _blobVersionedHash
-            )
+    /// @inheritdoc IScrollChain
+    function finalizeBundleWithProof(
+        bytes calldata _batchHeader,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _aggrProof
+    ) external override OnlyProver whenNotPaused {
+        if (_postStateRoot == bytes32(0)) revert ErrorStateRootIsZero();
+
+        // retrieve finalized state root and batch hash from storage
+        uint256 _finalizedBatchIndex = lastFinalizedBatchIndex;
+        bytes32 _prevStateRoot = finalizedStateRoots[_finalizedBatchIndex];
+        bytes32 _finalizedBatchHash = committedBatches[_finalizedBatchIndex];
+
+        // compute pending batch hash and verify
+        (uint256 batchPtr, bytes32 _batchHash, uint256 _batchIndex, ) = _loadBatchHeader(_batchHeader);
+        if (_batchIndex <= _finalizedBatchIndex) revert ErrorBatchIsAlreadyVerified();
```

Comment:
Is there any place to check if all batches before this bundle are finalized?

in the mock contract test (removing proof verification part), if rollup-relayer finalizes a bundle (e.g., containing batch 3), then previous batches (batch 2) cannot be finalized because the `lastFinalizedBatchIndex` is updated to 3.

Reply:
if a bundle containing batch 3 is finalized, we assume batch 2 is also finalized. So checking `lastFinalizedBatchIndex` is enough to check whether a batch is finalized or not.